### DISCUSSION TITLE
Usage question

### DISCUSSION THREAD

**Discussion by yuusheng** [ai/nanostores#156] [General]
Every time we use nanostores in our components we use `useStore` to ensure we subscribe to the reactivity inside nanostores.
But this kind of usage is not convenient enough. What if we write the code like this(Solid.js for example):

before
```js
// src/store/index.js
export const count = atom(0)

export function increase() {
  count.set(count.get() + 1)
}
export function decrease() {
  count.set(count.get() - 1)
}
```
```jsx
// src/Counter.jsx
import { count, increase, decrease } from './store'
import { useStore } from '@nanostores/solid'

export default function Counter() {
   const $count = useStore(count)

  return <div>
    <button onClick={increase}>+</button>
    <span>{ $count() }</span>
    <button onClick={decrease}>-</button>
  </div>
}
```

after
```js
// src/store/index.js
import { atom } from 'nanostores'
import { useStore } from '@nanostores/solid'

const count = atom(0)

export function useCounter() {
  const $count = useStore(count)

  function increase() {
    count.set(count.get() + 1)
  }

  function decrease() {
    count.set(count.get() - 1)
  }
  return {
    $count,
    increase,
    decrease,
  }
}
```

```jsx
// src/Counter.jsx
import { useCounter } from './store'

export default function Counter() {
  const { $count, increase, decrease } = useCounter()

  return <div>
    <button onClick={increase}>+</button>
    <span>{ $count() }</span>
    <button onClick={decrease}>-</button>
  </div>
}
```

Are there any performance issues or other problems?

**Comment by ai** [comment]
Both ways will be pretty similar in terms of performance.

The `useCounter()` has 2 small issues:

1. `increase()` and `decrease()` functions will be defined again on every component’s render.
2. If you do not use `decrease()` in the component, it will not be tree-shaked from JS bundle.

There are microoptimizations (especially in this use case). I personally prefer to define functions only once (rather than do it on every render), but it is subjective and will not be visible by user.

**Reply by yuusheng** [reply]
Thank you for your answer!
And for the first issue, solidjs render function will run only once at the beginning of execution. Does this still lead to the problem of redefinition?
And does using `useCallback` solve this problem for react?

**Reply by ai** [reply]
Yes, for Solid.js it will work much better.

> And does using useCallback solve this problem for react?

Nope. `useCallback` is also not cheap.
